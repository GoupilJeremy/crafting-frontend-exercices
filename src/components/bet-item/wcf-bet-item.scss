@import '../../shared/styles.scss';

button {
    @extend %bet-button;
}

.bet-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 10px 0;
    padding: 5px 10px;
    background: $bg-grey;
}

.bet-item__teams {
    display: flex;
    align-items: center;

    img {
        width: 35px;
        margin: 0 4px;
    }

    &--name {
        font-weight: $text-bold;
        margin: 0 8px;
    }
}

.bet-item__odds {
    &--name {
        display: block;
        width: 100%;
        margin: 5px 0;
        text-align: center;
        font-size: $small-text;
    }

    &--number {
        font-weight: $text-bold;
    }
}