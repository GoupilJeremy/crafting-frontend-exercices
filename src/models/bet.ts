import { GameInfo } from "./game";
import { OddsInfo } from "./odds";

export type BetInfo = Omit<GameInfo & OddsInfo, 'id'> & {
    [key: string]: number;
};

export type BetChoice = '1' | 'draw' | '2';
export type Bet = {
    gameId: string;
    selectedChoice: BetChoice;
    selectedOdd: number;
}