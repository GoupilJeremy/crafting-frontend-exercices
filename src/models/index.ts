export * from './games'
export * from './api'
export * from './bets'
export * from './storybook'