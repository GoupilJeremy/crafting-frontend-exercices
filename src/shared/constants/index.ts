export * from './events'
export * from './betChoice'