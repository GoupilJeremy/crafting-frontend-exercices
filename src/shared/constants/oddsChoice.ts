export const CHOICE_LEFT = '1';
export const CHOICE_DRAW = 'draw';
export const CHOICE_RIGHT = '2';