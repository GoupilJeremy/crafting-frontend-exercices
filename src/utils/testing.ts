export function mockPropsTo(element: Element) {
    return function (key: string, value: unknown) {
        element.setAttribute(key, JSON.stringify(value))
    }
}

export function dispatchMockedEventWith(element: Element) {
    return function (key: string, value: number | string) {
        const inputElement = element.shadowRoot?.querySelector('input')!;
        inputElement.value = value.toString();

        const fakeEvent = new Event(key)
        inputElement.dispatchEvent(fakeEvent);
    }
}

export function findElementWith(element: Element, key: string): Element | null | undefined {
    return element.shadowRoot?.querySelector(key);
}

export function isVisible(element: Element | null | undefined): boolean {
    if (!element) return false;

    return !element.hasAttribute('hidden');
}