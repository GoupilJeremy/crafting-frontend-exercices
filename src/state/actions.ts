import { BetChoice, BetInfo, User } from "../models";

export const UPDATE_BET_INFOS = 'UPDATE_BET_INFOS';
export const UPDATE_SELECTED_BET = 'UPDATE_SELECTED_BET';
export const UPDATE_STARTING_BET = 'UPDATE_STARTING_BET';
export const UPDATE_USER = 'UPDATE_USER';

export const doUpdateBetInfos = (betInfos: BetInfo[]) => ({
    type: UPDATE_BET_INFOS,
    payload: { betInfos }
})

export const doUpdateSelectedBet = (betInfo: BetInfo, choice: BetChoice) => ({
    type: UPDATE_SELECTED_BET,
    payload: { betInfo, choice }
})

export const doUpdateStartingBet = (startingBet: number) => ({
    type: UPDATE_STARTING_BET,
    payload: { startingBet }
})

export const doUpdateUser = (user: User) => ({
    type: UPDATE_USER,
    payload: { user }
})