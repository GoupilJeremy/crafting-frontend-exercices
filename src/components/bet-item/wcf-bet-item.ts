import css from './wcf-bet-item.scss';
import { BetChoice, BetInfo } from "../../models/bet";
import betIcon from '../../assets/bet.png'
import { CHOICE_1, CHOICE_2, CHOICE_DRAW } from '../../shared/constants/oddsChoice';
import { reduxStore } from '../../state/store';
import { doUpdateSelectedBet, UPDATE_SELECTED_BET } from '../../state/actions';

const template = document.createElement('template');
template.innerHTML = `
<style>${css}</style>

<div class="bet-item">
    <div class="bet-item__teams">
        <img src="${betIcon}" alt="Sport icon" />
        <p>
            <span class="bet-item__teams--name"></span> - 
            <span class="bet-item__teams--name"></span>
        </p>
    </div>

    <div class="bet-item__odds">
        <button>
            <span class="bet-item__odds--name"></span>
            <span class="bet-item__odds--number"></span>
        </button>
        <button>
            <span class="bet-item__odds--name">Draw</span>
            <span class="bet-item__odds--number"></span>
        </button>
        <button>
            <span class="bet-item__odds--name"></span>
            <span class="bet-item__odds--number"></span>
        </button>
    </div>
</div>
`;

export class BetItem extends HTMLElement {
    private _bet: BetInfo | undefined;

    constructor() {
        super();

        this.attachShadow({ mode: 'open' })
            .appendChild(template.content.cloneNode(true));

        const buttons = this.shadowRoot!.querySelectorAll('.bet-item__odds button');
        buttons[0].addEventListener('click', () => this.handleSelectBet(buttons[0], CHOICE_1));
        buttons[1].addEventListener('click', () => this.handleSelectBet(buttons[1], CHOICE_DRAW));
        buttons[2].addEventListener('click', () => this.handleSelectBet(buttons[2], CHOICE_2));
    }

    static get observedAttributes() {
        return ['bet']
    }

    attributeChangedCallback(name: string, _oldValue: string, newValue: string) {
        if (name !== 'bet') return;

        this.bet = JSON.parse(newValue);
        this.fillBetHTML();
    }

    get bet(): BetInfo {
        return this._bet!;
    }
    set bet(value: BetInfo) {
        this._bet = value
    }

    fillBetHTML() {
        const teamNames = this.shadowRoot!.querySelectorAll('.bet-item__teams--name')
        teamNames[0].textContent = this.bet.adversary1;
        teamNames[1].textContent = this.bet.adversary2;

        const oddsName = this.shadowRoot!.querySelectorAll('.bet-item__odds--name')
        oddsName[0].textContent = this.bet.adversary1;
        oddsName[2].textContent = this.bet.adversary2;

        const oddsNumber = this.shadowRoot!.querySelectorAll('.bet-item__odds--number')
        oddsNumber[0].textContent = this.bet.odd1.toFixed(2);
        oddsNumber[1].textContent = this.bet.odddraw.toFixed(2);
        oddsNumber[2].textContent = this.bet.odd2.toFixed(2);
    }

    selectClickedButton(clickedButton: Element) {
        const buttons = this.shadowRoot!.querySelectorAll('.bet-item__odds button');

        buttons.forEach((button: Element) => {
            if (button === clickedButton) {
                button.classList.add('selected')
            } else {
                button.classList.remove('selected')
            }
        })
    }

    handleSelectBet(buttonElement: Element, choice: BetChoice) {
        this.selectClickedButton(buttonElement);
        this.sentSelectedBet(this.bet, choice)
    }

    sentSelectedBet(betInfo: BetInfo, choice: BetChoice) {
        reduxStore.dispatch(doUpdateSelectedBet(betInfo, choice));
    }
}