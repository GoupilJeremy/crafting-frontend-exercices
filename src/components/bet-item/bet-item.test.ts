import { BetInfo } from "../../models/bet";
import { CHOICE_DRAW } from "../../shared/constants/oddsChoice";
import { doUpdateSelectedBet } from "../../state/actions";
import { reduxStore } from "../../state/store";
import { findElementWith } from "../../utils/testing";
import { BetItem } from "./bet-item";

const BET: BetInfo = {
    type: 'football',
    adversary1: 'Paris SG',
    adversary2: 'Lyon',
    gameId: '2',
    odd1: 1.12,
    odddraw: 2.50,
    odd2: 3.62
};

let betItem: BetItem = new BetItem();
betItem.setAttribute('bet', JSON.stringify(BET))

reduxStore.dispatch = jest.fn()

describe('BetItem Component', () => {
    it('should correctly render bet line', () => {
        expect(betItem?.shadowRoot?.querySelector('.bet-item')).toMatchSnapshot()
    });

    it('should send user choice when user click on a bet', () => {
        // Click on Draw
        clickOnButtonNumber(2);

        expect(reduxStore.dispatch).toHaveBeenCalledWith(doUpdateSelectedBet(BET, CHOICE_DRAW))
    });
});

function clickOnButtonNumber(nb: number) {
    const itemElement = findElementWith(betItem, '.bet-item')!
    const buttonElement = itemElement.querySelectorAll('button')[nb - 1]

    buttonElement.click();
}