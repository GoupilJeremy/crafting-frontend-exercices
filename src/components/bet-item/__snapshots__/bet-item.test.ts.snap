// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetItem Component should correctly render bet line 1`] = `
<div
  class="bet-item"
>
  
    
  <div
    class="bet-item__teams"
  >
    
        
    <img
      alt="Sport icon"
      src="test-file-stub"
    />
    
        
    <p>
      
            
      <span
        class="bet-item__teams--name"
      >
        Paris SG
      </span>
       - 
            
      <span
        class="bet-item__teams--name"
      >
        Lyon
      </span>
      
        
    </p>
    
    
  </div>
  

    
  <div
    class="bet-item__odds"
  >
    
        
    <button>
      
            
      <span
        class="bet-item__odds--name"
      >
        Paris SG
      </span>
      
            
      <span
        class="bet-item__odds--number"
      >
        1.12
      </span>
      
        
    </button>
    
        
    <button>
      
            
      <span
        class="bet-item__odds--name"
      >
        Draw
      </span>
      
            
      <span
        class="bet-item__odds--number"
      >
        2.50
      </span>
      
        
    </button>
    
        
    <button>
      
            
      <span
        class="bet-item__odds--name"
      >
        Lyon
      </span>
      
            
      <span
        class="bet-item__odds--number"
      >
        3.62
      </span>
      
        
    </button>
    
    
  </div>
  

</div>
`;
